skimage/feature/masked_register_translation.py

```diff
@@ -0,0 +1,275 @@
+
+    xcorr = mnxc(target_image, src_image, target_mask, src_mask,
+                 axes=(0, 1), mode='full', overlap_ratio=overlap_ratio)
+
+    # Generalize to the average of multiple equal maxima
+    maxima = np.transpose(np.nonzero(xcorr == xcorr.max()))
+    center = np.mean(maxima, axis=0)
+    shifts = center - np.array(src_image.shape) + 1
+    return -shifts + (size_mismatch / 2)
+
+
+def mnxc(arr1, arr2, m1, m2, mode='full', axes=(-2, -1), overlap_ratio=3 / 10):
+    """
+    Masked normalized cross-correlation (MNXC) between arrays.
+
+    Parameters
+    ----------
+    arr1 : ndarray
+        First array.
+    arr2 : ndarray
+        Seconds array. The dimensions of `arr2` along axes that are not
+        transformed should be equal to that of `arr1`.
+    m1 : ndarray
+        Mask of `arr1`. The mask should evaluate to `True`
+        (or 1) on valid pixels. `m1` should have the same shape as `arr1`.
+    m2 : ndarray
+        Mask of `arr2`. The mask should evaluate to `True`
+        (or 1) on valid pixels. `m2` should have the same shape as `arr2`.
+    mode : {'full', 'same'}, optional
+        'full':
+            This returns the convolution at each point of overlap. At
+            the end-points of the convolution, the signals do not overlap
+            completely, and boundary effects may be seen.
+        'same':
+            The output is the same size as `arr1`, centered with respect
+            to the `‘full’` output. Boundary effects are less prominent.
+    axes : tuple of ints, optional
+        Axes along which to compute the cross-correlation.
+    overlap_ratio : float, optional
+        Maximum allowed overlap ratio between masks. The correlation at pixels
+        with overlap ratio higher than this threshold will be zeroed.
+
+    Returns
+    -------
+    out : ndarray
+        Masked normalized cross-correlation.
+
+    Raises
+    ------
+    ValueError : if correlation `mode` is not valid, or array dimensions along
+        non-transformation axes are not equal.
+
+    References
+    ----------
+    .. [1] Dirk Padfield. Masked Object Registration in the Fourier Domain.
+        IEEE Transactions on Image Processing, vol.21(5), pp. 2706-2718 (2012).
+    """
+    if mode not in {'full', 'same'}:
+        raise ValueError("Correlation mode {} is not valid.".format(mode))
+
+    fixed_image = np.array(arr1, dtype=np.float)
+    fixed_mask = np.array(m1, dtype=np.bool)
+    moving_image = np.array(arr2, dtype=np.float)
+    moving_mask = np.array(m2, dtype=np.bool)
+    eps = np.finfo(np.float).eps
+
+    # Array dimensions along non-transformation axes should be equal.
+    all_axes = set(range(fixed_image.ndim))
+    for axis in (all_axes - set(axes)):
+        if fixed_image.shape[axis] != moving_image.shape[axis]:
+            raise ValueError(
+                'Array shapes along non-transformation axes should be \
+                    equal, but dimensions along axis {a} not'.format(a=axis))

```

Comment:
What are the two last characters on the two lines above?

Reply:
These were '\' for multiline strings. I thought they were necessary, but turns out that's not true! Removed.